queue.cpp

```diff
@@ -0,0 +1,99 @@
+#include <iostream>
+#include <stack>
+#include <algorithm>
+#include <functional>
+
+using namespace std;
+
+template<typename T, typename F = function<T(T, T)>>
+class AggQueue {
+private:
+    stack<pair<T, T>> inputStack, outputStack;
+    F func;
+
+    void pushWithAgg(stack<pair<T, T>>& stack, T value) {
+        T currentAgg = stack.empty() ? value : func(value, stack.top().second);
+        stack.push({value, currentAgg});
+    }
+
+    void transferIfNeeded() {
+        if (outputStack.empty()) {
+            while (!inputStack.empty()) {
+                T element = inputStack.top().first;
+                pushWithAgg(outputStack, element);
+                inputStack.pop();
+            }
+        }
+    }
+
+public:
+    AggQueue(F f) : func(f) {}
+
+    void enqueue(T value) {
+        pushWithAgg(inputStack, value);
+    }
+
+    void dequeue() {
+        if (inputStack.empty() && outputStack.empty()) {
+            cout << "Queue is empty! Cannot dequeue." << endl;
```

Comment:
Lets throw (C++ way), or lets return  `std::optional<T>` (Personally I like Result/Optional error handling approach more, but up to you)